File: source/ts/src/jyotish/bhava.ts
import { greenwichApparentSiderealDegrees } from '../astronomy/topocentric';
import { computeAyanamsa, dateToJulianDay } from '../astronomy/ayanamsa';
import { meanObliquity } from './planets';
import { computeNatalBasis, type NatalBasis } from './natalBasis';
import { resolveMasaName } from '../i18n/resolver';
import { normalize360, degToRad, radToDeg } from '../utils/angle';
import { validateLocation, validateDate } from '../utils/validation';
import { PanchangError } from '../types/errors';
import type { HouseSystem, BirthChartOptions } from '../types/options';
import type { GeoLocation } from '../types/location';
import type { BhavaChart, HouseInfo } from '../types/jyotish';

const TWO_PI = 2 * Math.PI;

/**
 * The 12 house cusps (bhavas): `'whole-sign'` (default), `'equal'`, `'placidus-kp'`.
 * @throws `PanchangError` ('CIRCUMPOLAR') when a `placidus-kp` intermediate cusp is circumpolar.
 */
export function computeBhava(
  birthDate: Date,
  location: GeoLocation,
  options: BirthChartOptions = {},
): BhavaChart {
  validateDate(birthDate);
  validateLocation(location);
  return bhavaFromBasis(
    computeNatalBasis(birthDate, location, options),
    options.houseSystem ?? 'whole-sign',
  );
}

/** @internal */
export function bhavaFromBasis(basis: NatalBasis, system: HouseSystem): BhavaChart {
  const { birthDate, location, ayanamsaType, lang, lagna } = basis;
  const ascSidereal = lagna.siderealLongitude;

  const lstDeg = normalize360(greenwichApparentSiderealDegrees(birthDate) + location.longitude);
  const T = (dateToJulianDay(birthDate) - 2451545.0) / 36525.0;
  const εRad = degToRad(meanObliquity(T));
  const mcTropical = computeMcTropical(lstDeg, εRad);
  const ayanamsa = computeAyanamsa(birthDate, ayanamsaType);
  const mcSidereal = normalize360(mcTropical - ayanamsa);

  const cuspLongitudes = buildCusps(
    system,
    ascSidereal,
    mcTropical,
    ayanamsa,
    lstDeg,
    location.latitude,
    εRad,
  );

  const houses: HouseInfo[] = cuspLongitudes.map((lon, i) => {
    const rashiIndex = Math.floor(lon / 30);
    return {
      house: i + 1,
      cuspLongitude: lon,
      rashi: { index: rashiIndex, name: resolveMasaName(rashiIndex, lang) },
      degreeInRashi: lon - rashiIndex * 30,
    };
  });

  return {
    system,
    houses,
    ascendantLongitude: ascSidereal,
    mcLongitude: mcSidereal,
  };
}

function buildCusps(
  system: HouseSystem,
  ascSidereal: number,
  mcTropical: number,
  ayanamsa: number,
  lstDeg: number,
  latitudeDeg: number,
  εRad: number,
): number[] {
  switch (system) {
    case 'whole-sign': {
      const ascRashiStart = Math.floor(ascSidereal / 30) * 30;
      const out: number[] = [];
      for (let i = 0; i < 12; i++) out.push(normalize360(ascRashiStart + i * 30));
      return out;
    }
    case 'equal': {
      const out: number[] = [];
      for (let i = 0; i < 12; i++) out.push(normalize360(ascSidereal + i * 30));
      return out;
    }
    case 'placidus-kp': {
      return placidusCusps(mcTropical, ayanamsa, lstDeg, latitudeDeg, εRad);
    }
  }
}

function computeMcTropical(lstDeg: number, εRad: number): number {
  const θ = degToRad(lstDeg);
  const mc = Math.atan2(Math.sin(θ), Math.cos(θ) * Math.cos(εRad));
  return normalize360(radToDeg(mc));
}

function placidusCusps(
  mcTropical: number,
  ayanamsa: number,
  lstDeg: number,
  latitudeDeg: number,
  εRad: number,
): number[] {
  const θRad = degToRad(lstDeg);
  const φRad = degToRad(latitudeDeg);

  const ascTropical = computeAscTropical(θRad, φRad, εRad);

  const cusp11Trop = solvePlacidus(11, θRad, φRad, εRad, normalize360(mcTropical + 30));
  const cusp12Trop = solvePlacidus(12, θRad, φRad, εRad, normalize360(mcTropical + 60));
  const cusp2Trop  = solvePlacidus(2,  θRad, φRad, εRad, normalize360(ascTropical + 30));
  const cusp3Trop  = solvePlacidus(3,  θRad, φRad, εRad, normalize360(ascTropical + 60));

  const trop = [
    ascTropical,
    cusp2Trop,
    cusp3Trop,
    normalize360(mcTropical + 180),
    normalize360(cusp11Trop + 180),
    normalize360(cusp12Trop + 180),
    normalize360(ascTropical + 180),
    normalize360(cusp2Trop + 180),
    normalize360(cusp3Trop + 180),
    mcTropical,
    cusp11Trop,
    cusp12Trop,
  ];

  return trop.map((t) => normalize360(t - ayanamsa));
}

function computeAscTropical(θRad: number, φRad: number, εRad: number): number {
  const num = Math.cos(θRad);
  const den = -Math.sin(εRad) * Math.tan(φRad) - Math.cos(εRad) * Math.sin(θRad);
  return normalize360(radToDeg(Math.atan2(num, den)));
}

function solvePlacidus(
  cusp: 2 | 3 | 11 | 12,
  θRad: number,
  φRad: number,
  εRad: number,
  initialGuessDeg: number,
): number {
  let λRad = degToRad(initialGuessDeg);
  const tanφ = Math.tan(φRad);
  const sinε = Math.sin(εRad);
  const cosε = Math.cos(εRad);
  const tolRad = degToRad(1e-7);

  for (let i = 0; i < 60; i++) {
    const sinλ = Math.sin(λRad);
    const cosλ = Math.cos(λRad);
    const α = Math.atan2(sinλ * cosε, cosλ);
    const δ = Math.asin(sinλ * sinε);

    const cosArg = -tanφ * Math.tan(δ);
    if (cosArg <= -1 || cosArg >= 1) {
      throw new PanchangError(
        `Placidus cusp ${cusp} undefined at latitude ${radToDeg(φRad).toFixed(2)}° (circumpolar). Use 'whole-sign' or 'equal'.`,
        'CIRCUMPOLAR',
      );
    }
    const SDA = Math.acos(cosArg);
    const SNA = Math.PI - SDA;

    let targetα: number;
    switch (cusp) {
      case 11: targetα = θRad + SDA / 3; break;
      case 12: targetα = θRad + 2 * SDA / 3; break;
      case 2:  targetα = θRad + SDA + SNA / 3; break;
      case 3:  targetα = θRad + SDA + 2 * SNA / 3; break;
    }

    let dα = targetα - α;
    dα = ((dα + Math.PI) % TWO_PI + TWO_PI) % TWO_PI - Math.PI;

    λRad += dα;
    if (Math.abs(dα) < tolRad) {
      return normalize360(radToDeg(λRad));
    }
  }
  throw new PanchangError(
    `Placidus cusp ${cusp} did not converge`,
    'PLACIDUS_DIVERGED',
  );
}

File: source/go/internal/jyotish/bhava.go
package jyotish

import (
	"math"

	"github.com/ishankgupta95/panchang/source/go/v5/internal/astronomy"
	"github.com/ishankgupta95/panchang/source/go/v5/internal/i18n"
	"github.com/ishankgupta95/panchang/source/go/v5/internal/jsnum"
	"github.com/ishankgupta95/panchang/source/go/v5/internal/utils"
	"github.com/ishankgupta95/panchang/source/go/v5/types"
)

const twoPi = 2 * jsnum.PI

func ComputeBhava(
	ctx *astronomy.EphemerisCtx,
	birthMs int64,
	location types.GeoLocation,
	options BirthChartOptions,
) (types.BhavaChart, error) {
	if err := utils.ValidateDate(birthMs); err != nil {
		return types.BhavaChart{}, err
	}
	if err := utils.ValidateLocation(location); err != nil {
		return types.BhavaChart{}, err
	}
	basis, err := ComputeNatalBasis(ctx, birthMs, location, options)
	if err != nil {
		return types.BhavaChart{}, err
	}
	return BhavaFromBasis(&basis, resolveHouseSystem(options.HouseSystem))
}

func resolveHouseSystem(s types.HouseSystem) types.HouseSystem {
	if s == "" {
		return types.HouseSystemWholeSign
	}
	return s
}

func BhavaFromBasis(basis *NatalBasis, system types.HouseSystem) (types.BhavaChart, error) {
	ascSidereal := basis.Lagna.SiderealLongitude

	lstDeg := utils.Normalize360(
		astronomy.GreenwichApparentSiderealDegrees(basis.Ctx, basis.BirthMs) + basis.Location.Longitude)
	t := (astronomy.DateToJulianDay(basis.BirthMs) - 2451545.0) / 36525.0
	epsRad := utils.DegToRad(MeanObliquity(t))
	mcTropical := computeMcTropical(lstDeg, epsRad)
	ayanamsa, err := astronomy.ComputeAyanamsa(basis.BirthMs, basis.AyanamsaType)
	if err != nil {
		return types.BhavaChart{}, err
	}
	mcSidereal := utils.Normalize360(mcTropical - ayanamsa)

	cuspLongitudes, err := buildCusps(
		system, ascSidereal, mcTropical, ayanamsa, lstDeg, basis.Location.Latitude, epsRad)
	if err != nil {
		return types.BhavaChart{}, err
	}

	houses := make([]types.HouseInfo, 0, len(cuspLongitudes))
	for i, lon := range cuspLongitudes {
		rashiIndex := int(math.Floor(lon / 30))
		houses = append(houses, types.HouseInfo{
			House:         i + 1,
			CuspLongitude: lon,
			Rashi:         types.RashiInfo{Index: rashiIndex, Name: i18n.ResolveMasaName(rashiIndex, basis.Lang)},
			DegreeInRashi: lon - float64(float64(rashiIndex)*30), // anti-FMA barrier
		})
	}

	return types.BhavaChart{
		System:             system,
		Houses:             houses,
		AscendantLongitude: ascSidereal,
		MCLongitude:        mcSidereal,
	}, nil
}

func buildCusps(
	system types.HouseSystem,
	ascSidereal, mcTropical, ayanamsa, lstDeg, latitudeDeg, epsRad float64,
) ([]float64, error) {
	switch system {
	case types.HouseSystemWholeSign:
		ascRashiStart := math.Floor(ascSidereal/30) * 30
		out := make([]float64, 0, 12)
		for i := 0; i < 12; i++ {
			out = append(out, utils.Normalize360(ascRashiStart+float64(float64(i)*30)))
		}
		return out, nil
	case types.HouseSystemEqual:
		out := make([]float64, 0, 12)
		for i := 0; i < 12; i++ {
			out = append(out, utils.Normalize360(ascSidereal+float64(float64(i)*30)))
		}
		return out, nil
	case types.HouseSystemPlacidusKP:
		return placidusCusps(mcTropical, ayanamsa, lstDeg, latitudeDeg, epsRad)
	}
	return nil, types.Codef(types.ErrInvalidInput,
		"unknown house system %q; expected one of whole-sign, equal, placidus-kp", string(system))
}

func computeMcTropical(lstDeg, epsRad float64) float64 {
	theta := utils.DegToRad(lstDeg)
	mc := math.Atan2(math.Sin(theta), math.Cos(theta)*math.Cos(epsRad))
	return utils.Normalize360(utils.RadToDeg(mc))
}

func placidusCusps(mcTropical, ayanamsa, lstDeg, latitudeDeg, epsRad float64) ([]float64, error) {
	thetaRad := utils.DegToRad(lstDeg)
	phiRad := utils.DegToRad(latitudeDeg)

	ascTropical := computeAscTropical(thetaRad, phiRad, epsRad)

	cusp11Trop, err := solvePlacidus(11, thetaRad, phiRad, epsRad, utils.Normalize360(mcTropical+30))
	if err != nil {
		return nil, err
	}
	cusp12Trop, err := solvePlacidus(12, thetaRad, phiRad, epsRad, utils.Normalize360(mcTropical+60))
	if err != nil {
		return nil, err
	}
	cusp2Trop, err := solvePlacidus(2, thetaRad, phiRad, epsRad, utils.Normalize360(ascTropical+30))
	if err != nil {
		return nil, err
	}
	cusp3Trop, err := solvePlacidus(3, thetaRad, phiRad, epsRad, utils.Normalize360(ascTropical+60))
	if err != nil {
		return nil, err
	}

	trop := [12]float64{
		ascTropical,
		cusp2Trop,
		cusp3Trop,
		utils.Normalize360(mcTropical + 180),
		utils.Normalize360(cusp11Trop + 180),
		utils.Normalize360(cusp12Trop + 180),
		utils.Normalize360(ascTropical + 180),
		utils.Normalize360(cusp2Trop + 180),
		utils.Normalize360(cusp3Trop + 180),
		mcTropical,
		cusp11Trop,
		cusp12Trop,
	}

	out := make([]float64, 0, 12)
	for _, t := range trop {
		out = append(out, utils.Normalize360(t-ayanamsa))
	}
	return out, nil
}

func computeAscTropical(thetaRad, phiRad, epsRad float64) float64 {
	num := math.Cos(thetaRad)
	den := -float64(math.Sin(epsRad)*math.Tan(phiRad)) - float64(math.Cos(epsRad)*math.Sin(thetaRad)) // FMA barriers
	return utils.Normalize360(utils.RadToDeg(math.Atan2(num, den)))
}

func solvePlacidus(cusp int, thetaRad, phiRad, epsRad, initialGuessDeg float64) (float64, error) {
	lambdaRad := utils.DegToRad(initialGuessDeg)
	tanPhi := math.Tan(phiRad)
	sinEps := math.Sin(epsRad)
	cosEps := math.Cos(epsRad)
	tolRad := utils.DegToRad(1e-7)

	for i := 0; i < 60; i++ {
		sinLambda := math.Sin(lambdaRad)
		cosLambda := math.Cos(lambdaRad)
		alpha := math.Atan2(sinLambda*cosEps, cosLambda)
		delta := math.Asin(sinLambda * sinEps)

		cosArg := -tanPhi * math.Tan(delta)
		if cosArg <= -1 || cosArg >= 1 {
			return 0, types.Codef(types.ErrCircumpolar,
				"Placidus cusp %d undefined at latitude %.2f° (circumpolar). Use 'whole-sign' or 'equal'.",
				cusp, utils.RadToDeg(phiRad))
		}
		sda := math.Acos(cosArg)
		sna := jsnum.PI - sda

		var targetAlpha float64
		switch cusp {
		case 11:
			targetAlpha = thetaRad + sda/3
		case 12:
			targetAlpha = thetaRad + 2*sda/3
		case 2:
			targetAlpha = thetaRad + sda + sna/3
		case 3:
			targetAlpha = thetaRad + sda + 2*sna/3
		default:
			return 0, types.Codef(types.ErrInvalidInput,
				"solvePlacidus: cusp %d is not one of 2, 3, 11, 12", cusp)
		}

		dAlpha := targetAlpha - alpha
		dAlpha = jsnum.Mod(jsnum.Mod(dAlpha+jsnum.PI, twoPi)+twoPi, twoPi) - jsnum.PI // jsnum.Mod, not math.Remainder

		lambdaRad += dAlpha
		if math.Abs(dAlpha) < tolRad {
			return utils.Normalize360(utils.RadToDeg(lambdaRad)), nil
		}
	}
	return 0, types.Codef(types.ErrPlacidusDiverged, "Placidus cusp %d did not converge", cusp)
}
